lib/ukstreambuf/include/uk/streambuf.h

```diff
@@ -0,0 +1,277 @@
+	struct uk_alloc *_a;
+#endif /* CONFIG_LIBUKALLOC */
+};
+
+/**
+ * Initializes a streambuf structure
+ *
+ * @param sb Reference to streambuf structure to initialize
+ * @param buf Pointer to memory area that should be used as buffer (required)
+ * @param buflen Size of the buffer memory area (minimum is `1`)
+ * @param cfg_flags Configuration flags (see `UK_STREAMBUF_C_*`)
+ */
+void uk_streambuf_init(struct uk_streambuf *sb, void *buf, __sz buflen,
+		       int cfg_flags);
+
+/**
+ * Resets a streambuf object and its corresponding buffer according to its
+ * configuration (see flags `UK_STREAMBUF_C_*`)
+ */
+void uk_streambuf_reset(struct uk_streambuf *sb);
+
+#if CONFIG_LIBUKALLOC
+/**
+ * Allocates and initializes a streambuf object together with
+ * a buffer.
+ *
+ * @param a Reference to an allocator that is used for the allocation
+ * @param buflen Target size of the buffer memory area (minimum is `1`)
+ * @param bufalign Alignment for buffer memory area
+ * @param cfg_flags Configuration flags (see `UK_STREAMBUF_C_*`)
+ * @return An allocated streambuf object with buffer or __NULL in case of errors
+ */
+struct uk_streambuf *uk_streambuf_alloc2(struct uk_alloc *a, __sz buflen,
+					 __sz bufalign, int cfg_flags);
+
+/** Shortcut for `uk_streambuf_alloc2()` without specifying an alignment */
+#define uk_streambuf_alloc(a, buflen, flags)				\
+	uk_streambuf_alloc2((a), (buflen), 1, (flags))
+
+/**
+ * Releases a streambuf object that was allocated with `uk_streambuf_alloc()`
+ * or `uk_streambuf_alloc2()`.
+ *
+ * @param sb Streambuf structure to release
+ */
+void uk_streambuf_free(struct uk_streambuf *sb);
+#endif /* CONFIG_LIBUKALLOC */
+
+/**
+ * Get a pointer to the buffer of a streambuf object.
+ *
+ * @param sb Streambuf object
+ * @return Pointer to corresponding buffer
+ */
+#define uk_streambuf_buf(sb)						\
+	((void *) (sb)->bufp)
+
+/**
+ * Current seek position
+ * NOTE: If `UK_STREAMBUF_C_TERMSHIFT` is configured, the seek position always
+ *       points to the last byte of the previous write, assuming it is a
+ *       termination byte (e.g., '\0'-terminating character of a string) so that
+ *       it will be overwritten with appending data.
+ *
+ * @param sb Streambuf object
+ * @return Current seek offset
+ */
+#define uk_streambuf_seek(sb)						\
+	 ((((sb)->flags & UK_STREAMBUF_C_TERMSHIFT) && ((sb)->seek > 0)) \
+	 ? (sb)->seek - 1 : (sb)->seek)
+
+/**
+ * Current write pointer (derived from seek position)
+ *
+ * @param sb Streambuf object
+ * @return Current write pointer
+ */
+#define uk_streambuf_wptr(sb)						\
+	((void *) ((sb)->bufp +	uk_streambuf_seek(sb)))
+
+/**
+ * Current length of data on the buffer of a streambuf
+ * HINT: If `UK_STREAMBUF_C_TERMSHIFT` is _not_ configured, the length is equal
+ *       to the seek offset.
+ *
+ * @param sb Streambuf object
+ * @return Current length on the corresponding buffer
+ */
+#define uk_streambuf_len(sb)						\
+	((sb)->seek)
```

Comment:
Why do we consider the termination byte to be part of the length of the data when `UK_STREAMBUF_C_TERMSHIFT` is configured?  
The `uk_streambuf_left` macro doesn't do this, for example. It reports the terminating byte as a free byte which can be overwritten. 

My intuition was that it should be similar to `strlen` when that option is selected and that `uk_streambuf_len(sb)` should just be equal to `uk_streambuf_seek(sb)`. Of course, tests would need to be updated accordingly. It's possible I'm missing something out and this isn't really an issue, though.

Reply:
Hi @eduardvintila, this is a good question. My idea behind was that a stream buffer is also working on binary data without any notion of termination symbols. I did consider `uk_streambuf_len(sb)` as the size of actual used data on the buffer - irrespective if there are termination symbols. If `UK_STREAMBUF_C_TERMSHIFT` is set, `uk_streambuf_seek(sb)` returns the number that you are asking for. In principle I consider that `uk_streambuf_len(sb)` has a relation to `uk_streambuf_buflen(sb)` while `uk_streambuf_seek(sb)` has a relation to `uk_streambuf_left(sb)`. I think, to avoid confusion, I would add additional notes to the function descriptions so that this is more clear. What do you think?

Reply:
@skuenzer Yep, sure, makes sense to me and a few notes on this would be really good to have  👍🏻